### ISSUE TITLE
esp_mqtt_client_publish blocking and causing task watchdog to fire (IDFGH-3052)

### ISSUE THREAD

**Issue by will-emmerson** [espressif/esp-idf#5077]
## Environment
- Development Kit: none
- Module or chip used: ESP32-WROOM-32
- IDF version: v4.2-dev-1034-ge599b794b
- Build System: idf.py
- Compiler version: xtensa-esp32-elf-gcc (crosstool-NG esp-2020r1) 8.2.0
- Operating System: Linux
- Using an IDE?: Clion
- Power Supply: external 3.3V

## Problem Description
`esp_mqtt_client_publish` appears to be blocking and this is causing the task watchdog to fire. This is only happening intermittently e.g. after several hours/days.

I am sending a fairly small amount of messages overall, probably 30-40 a minute but these tend to be grouped. I am replaying the same sequence every minute so it's not as if a sudden change in throughput is causing the problem. I am using less than 1% of cpu time according to `vTaskGetRunTimeStats`.

I am calling `esp_task_wdt_reset` before and after `esp_mqtt_client_publish` and `CONFIG_ESP_TASK_WDT_TIMEOUT_S` is at the default of 5.

Actually as usual as I'm just about to submit this issue I might just have seen the problem in the coredump below - I can see that `esp_transport_write` has a timeout of 10s, but where is this set?
`esp_transport_write (t=0x3ffe8710, buffer=0x3ffe9f06 <error: Cannot access memory at address 0x3ffe9f06>, len=324, timeout_ms=10000)`
Or should I just set the watchdog timeout to be longer than 10s?

All my tasks are running at priority=1. I am currently running a test with my mqtt task at priority 2 and the problem hasn't happened yet but it hasn't been running long, I don't know how much difference priority would make.

### Expected Behavior
`esp_mqtt_client_publish` shouldn't block for longer than `CONFIG_ESP_TASK_WDT_TIMEOUT_S`

### Actual Behavior
`esp_mqtt_client_publish` blocks for longer than `CONFIG_ESP_TASK_WDT_TIMEOUT_S`

## Other items if possible

sdkconfig.defaults:
```
# ethernet
CONFIG_ETH_ENABLED=y
CONFIG_ETH_USE_ESP32_EMAC=y
CONFIG_ETH_PHY_INTERFACE_RMII=y
CONFIG_ETH_RMII_CLK_OUTPUT=y
CONFIG_ETH_RMII_CLK_OUT_GPIO=17
CONFIG_ETH_DMA_BUFFER_SIZE=1600
CONFIG_ETH_USE_SPI_ETHERNET=n

# misc
CONFIG_ESPTOOLPY_FLASHSIZE_4MB=y
CONFIG_PARTITION_TABLE_CUSTOM=y
CONFIG_ESP32_ENABLE_COREDUMP_TO_FLASH=y
CONFIG_ESP32_CORE_DUMP_STACK_SIZE=4096

# factory reset
CONFIG_BOOTLOADER_FACTORY_RESET=y
CONFIG_BOOTLOADER_NUM_PIN_FACTORY_RESET=35
CONFIG_BOOTLOADER_OTA_DATA_ERASE=y
CONFIG_BOOTLOADER_DATA_FACTORY_RESET="nvs"

# debugging
CONFIG_ESP_SYSTEM_PANIC_PRINT_HALT=y
CONFIG_HEAP_POISONING_COMPREHENSIVE=y
CONFIG_FREERTOS_USE_TRACE_FACILITY=y
CONFIG_FREERTOS_USE_STATS_FORMATTING_FUNCTIONS=y
CONFIG_FREERTOS_GENERATE_RUN_TIME_STATS=y
CONFIG_LOG_TIMESTAMP_SOURCE_SYSTEM=y
CONFIG_COMPILER_STACK_CHECK=y
CONFIG_COMPILER_STACK_CHECK_MODE_NORM=y
CONFIG_COMPILER_WARN_WRITE_STRINGS=y

CONFIG_ESP_TASK_WDT_PANIC=y
```

coredump:
```
===============================================================
==================== ESP32 CORE DUMP START ====================

Crashed task handle: 0x3ffbf93c, name: 'IDLE0', GDB name: 'process 1073477948'

================== CURRENT THREAD REGISTERS ===================
exccause       0x1d (StoreProhibitedCause)
excvaddr       0x0
epc1           0x4018348a
epc2           0x0
epc3           0x0
epc4           0x0
epc5           0x0
epc6           0x0
epc7           0x0
eps2           0x60a20
eps3           0x0
eps4           0x0
eps5           0x0
eps6           0x0
eps7           0x40182483
pc             0x40088e8e       0x40088e8e <xt_unhandled_exception+10>
lbeg           0x4000c349       1073791817
lend           0x4000c36b       1073791851
lcount         0x0      0
sar            0x10     16
ps             0x60021  393249
threadptr      <unavailable>
br             <unavailable>
scompare1      <unavailable>
acclo          <unavailable>
acchi          <unavailable>
m0             <unavailable>
m1             <unavailable>
m2             <unavailable>
m3             <unavailable>
expstate       <unavailable>
f64r_lo        <unavailable>
f64r_hi        <unavailable>
f64s           <unavailable>
fcr            <unavailable>
fsr            <unavailable>
a0             0x80088648       -2146924984
a1             0x3ffb18c0       1073420480
a2             0x3ffb192a       1073420586
a3             0x3ffb1957       1073420631
a4             0xa      10
a5             0x1000   4096
a6             0x3ffb18a7       1073420455
a7             0x60a23  395811
a8             0x0      0
a9             0x1      1
a10            0x3ffb190e       1073420558
a11            0x3ffb190e       1073420558
a12            0x0      0
a13            0x3ffaf980       1073412480
a14            0x3ffb682c       1073440812
a15            0x0      0

==================== CURRENT THREAD STACK =====================
#0  0x40088e8e in xt_unhandled_exception (frame=0x3ffb192a) at /home/willemmerson/esp/esp-idf/components/esp_system/port/panic_handler.c:512
#1  0x40088648 in esp_system_abort (details=0x3ffb192a \"abort() was called at PC 0x400d695b on core 0\") at /home/willemmerson/esp/esp-idf/components/esp_system/system_api.c:68
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

======================== THREADS INFO =========================
  Id   Target Id         Frame 
* 1    process 1073477948 0x40088e8e in xt_unhandled_exception (frame=0x3ffb192a) at /home/willemmerson/esp/esp-idf/components/esp_system/port/panic_handler.c:512
  2    process 1073479876 0x4000bff0 in ?? ()
  3    process 1073579332 0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
  4    process 1073492528 0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
  5    process 1073583820 0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
  6    process 1073567948 0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
  7    process 1073482584 0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
  8    process 1073501504 0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
  9    process 1073497016 0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
  10   process 1073638660 0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
  11   process 1073506784 0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
  12   process 1073469580 0x40096920 in xQueueGenericReceive (xQueue=0x3ffbd410, pvBuffer=0x0, xTicksToWait=<optimized out>, xJustPeeking=0) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
  13   process 1073467948 0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
  14   process 1073510976 0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
  15   process 1073517032 0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
  16   process 1073555832 0x4000bff0 in ?? ()
  17   process 1073473600 0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
  18   process 1073592256 0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
  19   process 1073412916 0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
  20   process 1073636672 0x40096920 in xQueueGenericReceive (xQueue=0x3ffe6830, pvBuffer=0x0, xTicksToWait=<optimized out>, xJustPeeking=0) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39

==================== THREAD 1 (TCB: 0x3ffbf93c, name: 'IDLE0') =====================
#0  0x40088e8e in xt_unhandled_exception (frame=0x3ffb192a) at /home/willemmerson/esp/esp-idf/components/esp_system/port/panic_handler.c:512
#1  0x40088648 in esp_system_abort (details=0x3ffb192a \"abort() was called at PC 0x400d695b on core 0\") at /home/willemmerson/esp/esp-idf/components/esp_system/system_api.c:68
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

==================== THREAD 2 (TCB: 0x3ffc00c4, name: 'IDLE1') =====================
#0  0x4000bff0 in ?? ()
#1  0x40095c50 in vPortExitCritical (mux=<optimized out>) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:419
#2  0x400d6b83 in esp_task_wdt_reset () at /home/willemmerson/esp/esp-idf/components/esp_common/src/task_wdt.c:320
#3  0x400d6b8e in idle_hook_cb () at /home/willemmerson/esp/esp-idf/components/esp_common/src/task_wdt.c:86
#4  0x400d6db0 in esp_vApplicationIdleHook () at /home/willemmerson/esp/esp-idf/components/esp_common/src/freertos_hooks.c:51
#5  0x40096edc in prvIdleTask (pvParameters=0x0) at /home/willemmerson/esp/esp-idf/components/freertos/tasks.c:3385
#6  0x400959ed in vPortTaskWrapper (pxCode=0x40096ed0 <prvIdleTask>, pvParameters=0x0) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 3 (TCB: 0x3ffd8544, name: 'uart_listen') =====================
#0  0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
#1  0x40096920 in xQueueGenericReceive (xQueue=0x3ffd5d70, pvBuffer=0x3ffd8080, xTicksToWait=<optimized out>, xJustPeeking=0) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#2  0x400d8a13 in coordinator_listen_task () at ../main/uart.c:157
#3  0x400959ed in vPortTaskWrapper (pxCode=0x400d89b8 <coordinator_listen_task>, pvParameters=0x0) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 4 (TCB: 0x3ffc3230, name: 'led_blink_task') =====================
#0  0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
#1  0x40098246 in vTaskDelay (xTicksToDelay=<optimized out>) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#2  0x400d7c3f in led_blink_task () at ../main/multiplexer.c:115
#3  0x400959ed in vPortTaskWrapper (pxCode=0x400d7c04 <led_blink_task>, pvParameters=0x0) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 5 (TCB: 0x3ffd96cc, name: 'status_task') =====================
#0  0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
#1  0x40098246 in vTaskDelay (xTicksToDelay=<optimized out>) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#2  0x400d8713 in status_task () at ../main/status.c:58
#3  0x400959ed in vPortTaskWrapper (pxCode=0x400d86cc <status_task>, pvParameters=0x0) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 6 (TCB: 0x3ffd58cc, name: 'powermeter') =====================
#0  0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
#1  0x400981dd in vTaskDelayUntil (pxPreviousWakeTime=<optimized out>, xTimeIncrement=<optimized out>) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#2  0x400d81f0 in powermeter_task () at ../main/powermeter.c:215
#3  0x400959ed in vPortTaskWrapper (pxCode=0x400d814c <powermeter_task>, pvParameters=0x0) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 7 (TCB: 0x3ffc0b58, name: 'Tmr Svc') =====================
#0  0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
#1  0x40098ecd in prvProcessTimerOrBlockTask (xNextExpireTime=<optimized out>, xListWasEmpty=<optimized out>) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#2  0x40098fcb in prvTimerTask (pvParameters=0x0) at /home/willemmerson/esp/esp-idf/components/freertos/timers.c:544
#3  0x400959ed in vPortTaskWrapper (pxCode=0x40098fbc <prvTimerTask>, pvParameters=0x0) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 8 (TCB: 0x3ffc5540, name: 'incoming_msgs') =====================
#0  0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
#1  0x40096920 in xQueueGenericReceive (xQueue=0x3ffc0d90, pvBuffer=0x3ffc547c, xTicksToWait=<optimized out>, xJustPeeking=0) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#2  0x400d9c56 in incoming_instruction_task () at ../main/instructions.c:78
#3  0x400959ed in vPortTaskWrapper (pxCode=0x400d9c18 <incoming_instruction_task>, pvParameters=0x0) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 9 (TCB: 0x3ffc43b8, name: 'relay_on_off') =====================
#0  0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
#1  0x40098246 in vTaskDelay (xTicksToDelay=<optimized out>) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#2  0x400d9f95 in relay_instruction_task () at ../main/instructions.c:179
#3  0x400959ed in vPortTaskWrapper (pxCode=0x400d9ec4 <relay_instruction_task>, pvParameters=0x0) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 10 (TCB: 0x3ffe6d04, name: 'mqtt_publish') =====================
#0  0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
#1  0x40096920 in xQueueGenericReceive (xQueue=0x3fff367c, pvBuffer=0x0, xTicksToWait=<optimized out>, xJustPeeking=0) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#2  0x401048b5 in sys_arch_sem_wait (sem=0x3ffd9b64, timeout=10000) at /home/willemmerson/esp/esp-idf/components/lwip/port/esp32/freertos/sys_arch.c:184
#3  0x401077a7 in lwip_select (maxfdp1=58, readset=0x0, writeset=0x3ffe7c80, exceptset=0x3ffe7c88, timeout=0x3ffe7c78) at /home/willemmerson/esp/esp-idf/components/lwip/lwip/src/api/sockets.c:2154
#4  0x400d409e in esp_vfs_select (nfds=58, readfds=0x0, writefds=0x3ffe7c80, errorfds=0x3ffe7c88, timeout=0x3ffe7c78) at /home/willemmerson/esp/esp-idf/components/vfs/vfs.c:985
#5  0x4011694e in ssl_poll_write (t=<optimized out>, timeout_ms=10000) at /home/willemmerson/esp/esp-idf/components/tcp_transport/transport_ssl.c:119
#6  0x40183610 in esp_transport_poll_write (t=0x3ffe864c, timeout_ms=10000) at /home/willemmerson/esp/esp-idf/components/tcp_transport/transport.c:206
#7  0x40116af7 in ssl_write (t=0x3ffe864c, buffer=0x3ffe9f06 <error: Cannot access memory at address 0x3ffe9f06>, len=324, timeout_ms=10000) at /home/willemmerson/esp/esp-idf/components/tcp_transport/transport_ssl.c:135
#8  0x401835d0 in esp_transport_write (t=0x3ffe864c, buffer=0x3ffe9f06 <error: Cannot access memory at address 0x3ffe9f06>, len=324, timeout_ms=10000) at /home/willemmerson/esp/esp-idf/components/tcp_transport/transport.c:190
#9  0x40117354 in _ws_write (t=<optimized out>, opcode=<optimized out>, mask_flag=128, b=0x3ffe9f06 <error: Cannot access memory at address 0x3ffe9f06>, len=324, timeout_ms=10000) at /home/willemmerson/esp/esp-idf/components/tcp_transport/transport_ws.c:257
#10 0x401173d3 in ws_write (t=0x3ffe8710, b=0x3ffe9f06 <error: Cannot access memory at address 0x3ffe9f06>, len=324, timeout_ms=10000) at /home/willemmerson/esp/esp-idf/components/tcp_transport/transport_ws.c:289
#11 0x401835d0 in esp_transport_write (t=0x3ffe8710, buffer=0x3ffe9f06 <error: Cannot access memory at address 0x3ffe9f06>, len=324, timeout_ms=10000) at /home/willemmerson/esp/esp-idf/components/tcp_transport/transport.c:190
#12 0x400e2309 in mqtt_write_data (client=0x3ffe7ff0) at /home/willemmerson/esp/esp-idf/components/mqtt/esp-mqtt/mqtt_client.c:773
#13 0x400e39dc in esp_mqtt_client_publish (client=0x3ffe7ff0, topic=<optimized out>, data=<optimized out>, len=<optimized out>, qos=0, retain=<optimized out>) at /home/willemmerson/esp/esp-idf/components/mqtt/esp-mqtt/mqtt_client.c:1576
#14 0x400d78ad in mqtt_publish (topic=0x3ffd476c <error: Cannot access memory at address 0x3ffd476c>, message=0x3fff3378 <error: Cannot access memory at address 0x3fff3378>, qos=0 '\\000') at ../main/mqtt.c:81
#15 0x400d7949 in mqtt_publish_task () at ../main/mqtt.c:101
#16 0x400959ed in vPortTaskWrapper (pxCode=0x400d78b4 <mqtt_publish_task>, pvParameters=0x0) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 11 (TCB: 0x3ffc69e0, name: 'tiT') =====================
#0  0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
#1  0x40096920 in xQueueGenericReceive (xQueue=0x3ffc5a94, pvBuffer=0x3ffc68cc, xTicksToWait=<optimized out>, xJustPeeking=0) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#2  0x401049a4 in sys_arch_mbox_fetch (mbox=<optimized out>, msg=0x3ffc68cc, timeout=40) at /home/willemmerson/esp/esp-idf/components/lwip/port/esp32/freertos/sys_arch.c:325
#3  0x400f1d53 in tcpip_timeouts_mbox_fetch (mbox=0x3ffb78ec <tcpip_mbox>, msg=0x3ffc68cc) at /home/willemmerson/esp/esp-idf/components/lwip/lwip/src/api/tcpip.c:110
#4  0x400f1e0a in tcpip_thread (arg=0x0) at /home/willemmerson/esp/esp-idf/components/lwip/lwip/src/api/tcpip.c:148
#5  0x400959ed in vPortTaskWrapper (pxCode=0x400f1df0 <tcpip_thread>, pvParameters=0x0) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 12 (TCB: 0x3ffbd88c, name: 'ipc1') =====================
#0  0x40096920 in xQueueGenericReceive (xQueue=0x3ffbd410, pvBuffer=0x0, xTicksToWait=<optimized out>, xJustPeeking=0) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#1  0x4008206f in ipc_task (arg=0x1) at /home/willemmerson/esp/esp-idf/components/esp_common/src/ipc.c:51
#2  0x400959ed in vPortTaskWrapper (pxCode=0x4008203c <ipc_task>, pvParameters=0x1) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 13 (TCB: 0x3ffbd22c, name: 'ipc0') =====================
#0  0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
#1  0x40096920 in xQueueGenericReceive (xQueue=0x3ffaff00, pvBuffer=0x0, xTicksToWait=<optimized out>, xJustPeeking=0) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#2  0x4008206f in ipc_task (arg=0x0) at /home/willemmerson/esp/esp-idf/components/esp_common/src/ipc.c:51
#3  0x400959ed in vPortTaskWrapper (pxCode=0x4008203c <ipc_task>, pvParameters=0x0) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 14 (TCB: 0x3ffc7a40, name: 'sys_evt') =====================
#0  0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
#1  0x40096920 in xQueueGenericReceive (xQueue=0x3ffc6c58, pvBuffer=0x3ffc7950, xTicksToWait=<optimized out>, xJustPeeking=0) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#2  0x4018622d in esp_event_loop_run (event_loop=0x3ffc6c30, ticks_to_run=4294967295) at /home/willemmerson/esp/esp-idf/components/esp_event/esp_event.c:622
#3  0x40186244 in esp_event_loop_run_task (args=0x3ffc6c30) at /home/willemmerson/esp/esp-idf/components/esp_event/esp_event.c:115
#4  0x400959ed in vPortTaskWrapper (pxCode=0x40186238 <esp_event_loop_run_task>, pvParameters=0x3ffc6c30) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 15 (TCB: 0x3ffc91e8, name: 'mdns') =====================
#0  0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
#1  0x40096920 in xQueueGenericReceive (xQueue=0x3ffc7e6c, pvBuffer=0x3ffc912c, xTicksToWait=<optimized out>, xJustPeeking=0) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#2  0x400dff9b in _mdns_service_task (pvParameters=<optimized out>) at /home/willemmerson/esp/esp-idf/components/mdns/mdns.c:4079
#3  0x400959ed in vPortTaskWrapper (pxCode=0x400dff7c <_mdns_service_task>, pvParameters=0x0) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 16 (TCB: 0x3ffd2978, name: 'emac_rx') =====================
#0  0x4000bff0 in ?? ()
#1  0x40095c50 in vPortExitCritical (mux=<optimized out>) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:419
#2  0x40098969 in ulTaskNotifyTake (xClearCountOnExit=0, xTicksToWait=<optimized out>) at /home/willemmerson/esp/esp-idf/components/freertos/tasks.c:4648
#3  0x400e7c49 in emac_esp32_rx_task (arg=0x3ffc9784) at /home/willemmerson/esp/esp-idf/components/esp_eth/src/esp_eth_mac_esp32.c:252
#4  0x400959ed in vPortTaskWrapper (pxCode=0x400e7c38 <emac_esp32_rx_task>, pvParameters=0x3ffc9784) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 17 (TCB: 0x3ffbe840, name: 'mqtt_task') =====================
#0  0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
#1  0x40096920 in xQueueGenericReceive (xQueue=0x3ffe80f0, pvBuffer=0x0, xTicksToWait=<optimized out>, xJustPeeking=0) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#2  0x4009697c in xQueueTakeMutexRecursive (xMutex=0x3ffe80f0, xTicksToWait=4294967295) at /home/willemmerson/esp/esp-idf/components/freertos/queue.c:634
#3  0x400e2806 in esp_mqtt_task (pv=0x3ffe7ff0) at /home/willemmerson/esp/esp-idf/components/mqtt/esp-mqtt/mqtt_client.c:1219
#4  0x400959ed in vPortTaskWrapper (pxCode=0x400e2798 <esp_mqtt_task>, pvParameters=0x3ffe7ff0) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 18 (TCB: 0x3ffdb7c0, name: 'wifi') =====================
#0  0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
#1  0x40096920 in xQueueGenericReceive (xQueue=0x3ffda304, pvBuffer=0x3ffdb6f0, xTicksToWait=<optimized out>, xJustPeeking=0) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#2  0x400f094c in queue_recv_wrapper (queue=0x3ffda304, item=0x3ffdb6f0, block_time_tick=4294967295) at /home/willemmerson/esp/esp-idf/components/esp_wifi/esp32/esp_adapter.c:345
#3  0x4008d53c in wdevProcessRxSucDataAll () at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#4  0x400959ed in vPortTaskWrapper (pxCode=0x4008d508 <wdevProcessRxSucDataAll+1412>, pvParameters=0x0) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 19 (TCB: 0x3ffafb34, name: 'esp_timer') =====================
#0  0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
#1  0x40096920 in xQueueGenericReceive (xQueue=0x3ffaeab8, pvBuffer=0x0, xTicksToWait=<optimized out>, xJustPeeking=0) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#2  0x400d6f4c in timer_task (arg=0x0) at /home/willemmerson/esp/esp-idf/components/esp_timer/src/esp_timer.c:325
#3  0x400959ed in vPortTaskWrapper (pxCode=0x400d6f38 <timer_task>, pvParameters=0x0) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143

==================== THREAD 20 (TCB: 0x3ffe6540, name: 'httpd') =====================
#0  0x40096920 in xQueueGenericReceive (xQueue=0x3ffe6830, pvBuffer=0x0, xTicksToWait=<optimized out>, xJustPeeking=0) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
#1  0x4010488e in sys_arch_sem_wait (sem=0x3ffe681c, timeout=0) at /home/willemmerson/esp/esp-idf/components/lwip/port/esp32/freertos/sys_arch.c:180
#2  0x401077a7 in lwip_select (maxfdp1=56, readset=0x3ffe6458, writeset=0x0, exceptset=0x0, timeout=0x0) at /home/willemmerson/esp/esp-idf/components/lwip/lwip/src/api/sockets.c:2154
#3  0x400d409e in esp_vfs_select (nfds=56, readfds=0x3ffe6458, writefds=0x0, errorfds=0x0, timeout=0x0) at /home/willemmerson/esp/esp-idf/components/vfs/vfs.c:985
#4  0x401198d1 in httpd_server (hd=0x3ffe4374) at /home/willemmerson/esp/esp-idf/components/esp_http_server/src/httpd_main.c:174
#5  0x401199d2 in httpd_thread (arg=0x3ffe4374) at /home/willemmerson/esp/esp-idf/components/esp_http_server/src/httpd_main.c:227
#6  0x400959ed in vPortTaskWrapper (pxCode=0x401199c4 <httpd_thread>, pvParameters=0x3ffe4374) at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/port.c:143


======================= ALL MEMORY REGIONS ========================
Name   Address   Size   Attrs
.rtc.text 0x400c0000 0x0 RW  
.rtc.dummy 0x3ff80000 0x0 RW  
.rtc.force_fast 0x3ff80000 0x0 RW  
.rtc_noinit 0x50000000 0x0 RW  
.rtc.force_slow 0x50000000 0x0 RW  
.iram0.vectors 0x40080000 0x404 R XA
.iram0.text 0x40080404 0x19128 R XA
.dram0.data 0x3ffb0000 0x4f70 RW A
.noinit 0x3ffb4f70 0x0 RW  
.flash.rodata 0x3f400020 0x2cb98 RW A
.flash.text 0x400d0020 0xb6b24 R XA
.iram0.text_end 0x4009952c 0x0 RW  
.dram0.heap_start 0x3ffbcdf8 0x0 RW  
.coredump.tasks.data 0x3ffbf780 0x1a8 RW 
.coredump.tasks.data 0x3ffc00c4 0x168 RW 
.coredump.tasks.data 0x3ffbfeb0 0x200 RW 
.coredump.tasks.data 0x3ffbf93c 0x168 RW 
.coredump.tasks.data 0x3ffd8544 0x168 RW 
.coredump.tasks.data 0x3ffd7f40 0x5f0 RW 
.coredump.tasks.data 0x3ffc3230 0x168 RW 
.coredump.tasks.data 0x3ffc3070 0x1ac RW 
.coredump.tasks.data 0x3ffd96cc 0x168 RW 
.coredump.tasks.data 0x3ffd9510 0x1a8 RW 
.coredump.tasks.data 0x3ffd58cc 0x168 RW 
.coredump.tasks.data 0x3ffd56f0 0x1c8 RW 
.coredump.tasks.data 0x3ffc0b58 0x168 RW 
.coredump.tasks.data 0x3ffc0980 0x1c4 RW 
.coredump.tasks.data 0x3ffc5540 0x168 RW 
.coredump.tasks.data 0x3ffc5350 0x1dc RW 
.coredump.tasks.data 0x3ffc43b8 0x168 RW 
.coredump.tasks.data 0x3ffc41f0 0x1b4 RW 
.coredump.tasks.data 0x3ffe6d04 0x168 RW 
.coredump.tasks.data 0x3ffe7a10 0x4ec RW 
.coredump.tasks.data 0x3ffc69e0 0x168 RW 
.coredump.tasks.data 0x3ffc6750 0x22c RW 
.coredump.tasks.data 0x3ffbd88c 0x168 RW 
.coredump.tasks.data 0x3ffbd6d0 0x1a8 RW 
.coredump.tasks.data 0x3ffbd22c 0x168 RW 
.coredump.tasks.data 0x3ffbd050 0x1c8 RW 
.coredump.tasks.data 0x3ffc7a40 0x168 RW 
.coredump.tasks.data 0x3ffc7820 0x20c RW 
.coredump.tasks.data 0x3ffc91e8 0x168 RW 
.coredump.tasks.data 0x3ffc9000 0x1d4 RW 
.coredump.tasks.data 0x3ffd2978 0x168 RW 
.coredump.tasks.data 0x3ffd27a0 0x1c4 RW 
.coredump.tasks.data 0x3ffbe840 0x168 RW 
.coredump.tasks.data 0x3ffec8d0 0x1f0 RW 
.coredump.tasks.data 0x3ffdb7c0 0x168 RW 
.coredump.tasks.data 0x3ffdb5b0 0x1fc RW 
.coredump.tasks.data 0x3ffafb34 0x168 RW 
.coredump.tasks.data 0x3ffaf960 0x1c0 RW 
.coredump.tasks.data 0x3ffe6540 0x168 RW 
.coredump.tasks.data 0x3ffe6210 0x31c RW 

===================== ESP32 CORE DUMP END =====================
===============================================================

```

**Comment by Alvin1Zhang** [comment]
@will-emmerson Thanks for the detailed report, we will look into.

**Comment by david-cermak** [comment]
Hi @will-emmerson 

As for the `esp_transport_write()` there's no need for setting WTD timeout to the transport timeout (10s) or higher, since it internally uses select, so even if it blocks for the whole 10s it still yields to lower prio tasks if socket not writable (and this is what's seen in the core dump, btw)

The transport timeout is not adjustable at the moment and is only set here https://github.com/espressif/esp-mqtt/blob/master/include/mqtt_config.h#L45

It might be possible that the `esp_mqtt_client_publish()` would "busy block" (i.e. causing the IDLE tasks to starve and the WDT to trigger), but that would practically happen only and when publishing very large data (typically MBs), as the payloads longer than the mqtt buffer would get fragmented and transmitted in a loop. What is the approximate size of your data to publish, after they accumulate?
But, this is not very likely to happen and, looking at your core dump, this is not what's happing here:
```
==================== THREAD 10 (TCB: 0x3ffe6d04, name: 'mqtt_publish') =====================
#0  0x40081a38 in esp_crosscore_int_send_yield (core_id=0) at /home/willemmerson/esp/esp-idf/components/esp32/crosscore_int.c:110
#1  0x40096920 in xQueueGenericReceive (xQueue=0x3fff367c, pvBuffer=0x0, xTicksToWait=<optimized out>, xJustPeeking=0) at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:39
...
#3  0x401077a7 in lwip_select (maxfdp1=58, readset=0x0, writeset=0x3ffe7c80, exceptset=0x3ffe7c88, timeout=0x3ffe7c78) at /home/willemmerson/esp/esp-idf/components/lwip/lwip/src/api/sockets.c:2154
...
#13 0x400e39dc in esp_mqtt_client_publish (client=0x3ffe7ff0, topic=<optimized out>, data=<optimized out>, len=<optimized out>, qos=0, retain=<optimized out>) at /home/willemmerson/esp/esp-idf/components/mqtt/esp-mqtt/mqtt_client.c:1576
```
as you can see, the `esp_mqtt_client_publish()` in this case is blocked on select.

It is not very clear from this core dump what is the source of the abort() in IDLE task. I can see that all listed tasks are generally blocked on some semaphore/queue or similar. One potential issue is that the core dump is not complete, which might be the case if both cores are involved. What would help in this case is, if you can provide the actual backtrace from the board log, so something like:
```
Backtrace:0x400D1F04:0x3FFB0750 0x400821DD:0x3FFB0770 0x400D29A1:0x3FFB4810 0x400D149E:0x3FFB4840
```
with that information and knowing the elf file for decoding the addresses we could better understand the source of the issue.

**Comment by will-emmerson** [comment]
Unfortunately I don't have the original coredump so will have to try and recreate the problem again. The coredump was saved to flash, sent to a server on the next boot and then was viewed with `espcoredump.py info_corefile build/8-channel.elf --core coredump.elf` so maybe something was lost in that process.
I set `CONFIG_TASK_WDT_TIMEOUT_S=15` and the problem hasn't happened since then (been running fine for 10 days or so) which supports the idea it's something to do with mqtt timeout but then again, could be anything.
I am not sending anything large, just tiny messages of a few hundred chars, and not really that many.

I am confused about the task watchdog, because even tasks which yield to lower priority tasks cause the watchdog to fire. For instance in this code, commenting out `esp_task_wdt_add` makes watchdog fire even though `vTaskDelay` should be yielding:

```
static void led_blink_task() {
    ESP_ERROR_CHECK(esp_task_wdt_add(NULL));
    while (true) {
//        ESP_ERROR_CHECK(esp_task_wdt_reset());
        pin_toggle(LED1);
        vTaskDelay(ONE_SECOND);
    }
}
```

**Comment by will-emmerson** [comment]
Is there something weird going on with `emac_rx` task in the above coredump by the way? It's the only task not waiting.

**Comment by david-cermak** [comment]
> because even tasks which yield to lower priority tasks cause the watchdog to fire

It should not, if you start from a clean project and setup a very high prio task to
``` c
while(1) {
    vTaskDelay(ONE_YEAR);
    }
```
the task WDT should *not* fire. Does the error message indicate which task/core caused the WDT expiry? was it the `led_blink_task()`?

> Is there something weird going on with emac_rx task in the above coredump

I don't think so, it's blocked on a `ulTaskNotifyTake()` which is also a kind of a semaphore...

It would be interesting to see what's happening on the other core, but unfortunately the core dump doesn't list tasks on both cores (there is a plan to support this soon in IDF)

There's been another issue https://github.com/espressif/esp-idf/issues/1981#issuecomment-516743396, which might indicate there could be something wrong with the task WDT triggering and consequent processing, but has to do with UART interrupts/not clearing flags...

**Comment by will-emmerson** [comment]
I did as you said and created a clean project with the following code but the task watchdog still fired:

```c
#include <esp_task_wdt.h>

void loop_task() {
    ESP_ERROR_CHECK(esp_task_wdt_add(NULL));
    while (true) {
        vTaskDelay(portMAX_DELAY);
    }
}

void app_main(void) {
    xTaskCreate(loop_task, "loop_task", 2048, NULL, configMAX_PRIORITIES, NULL);
}
```

Output:

```
�ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4
load:0x3fff0034,len:7148
load:0x40078000,len:13696
load:0x40080400,len:3984
0x40080400: _init at ??:?

entry 0x40080688
I (29) boot: ESP-IDF v4.2-dev-1034-ge599b794b-dirty 2nd stage bootloader
I (29) boot: compile time 09:25:35
I (30) boot: chip revision: 1
I (34) boot_comm: chip revision: 1, min. bootloader chip revision: 0
I (41) boot.esp32: SPI Speed      : 40MHz
I (45) boot.esp32: SPI Mode       : DIO
I (50) boot.esp32: SPI Flash Size : 2MB
I (54) boot: Enabling RNG early entropy source...
I (60) boot: Partition Table:
I (63) boot: ## Label            Usage          Type ST Offset   Length
I (71) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (78) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (86) boot:  2 factory          factory app      00 00 00010000 00100000
I (93) boot: End of partition table
I (97) boot_comm: chip revision: 1, min. application chip revision: 0
I (104) esp_image: segment 0: paddr=0x00010020 vaddr=0x3f400020 size=0x053e8 ( 21480) map
I (122) esp_image: segment 1: paddr=0x00015410 vaddr=0x3ffb0000 size=0x0225c (  8796) load
I (126) esp_image: segment 2: paddr=0x00017674 vaddr=0x40080000 size=0x00404 (  1028) load
0x40080000: _WindowOverflow4 at /home/willemmerson/esp/esp-idf/components/freertos/xtensa/xtensa_vectors.S:1730

I (132) esp_image: segment 3: paddr=0x00017a80 vaddr=0x40080404 size=0x08598 ( 34200) load
I (155) esp_image: segment 4: paddr=0x00020020 vaddr=0x400d0020 size=0x12878 ( 75896) map
0x400d0020: _stext at ??:?

I (184) esp_image: segment 5: paddr=0x000328a0 vaddr=0x4008899c size=0x016a0 (  5792) load
0x4008899c: cpu_ll_get_core_id at /home/willemmerson/esp/esp-idf/components/soc/src/esp32/include/hal/cpu_ll.h:35
 (inlined by) xPortGetCoreID at /home/willemmerson/esp/esp-idf/components/freertos/include/freertos/portable.h:205
 (inlined by) prvAddNewTaskToReadyList at /home/willemmerson/esp/esp-idf/components/freertos/tasks.c:1185

I (193) boot: Loaded app from partition at offset 0x10000
I (193) boot: Disabling RNG early entropy source...
I (194) cpu_start: Pro cpu up.
I (197) cpu_start: Application information:
I (202) cpu_start: Project name:     empty_esp32_project
I (208) cpu_start: App version:      1
I (212) cpu_start: Compile time:     Apr 20 2020 09:25:26
I (219) cpu_start: ELF file SHA256:  f8f5f51bbd9e2d21...
I (225) cpu_start: ESP-IDF:          v4.2-dev-1034-ge599b794b-dirty
I (231) cpu_start: Starting app cpu, entry point is 0x400811f8
0x400811f8: call_start_cpu1 at /home/willemmerson/esp/esp-idf/components/esp32/cpu_start.c:277

I (0) cpu_start: App cpu up.
I (242) heap_init: Initializing. RAM available for dynamic allocation:
I (249) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM
I (255) heap_init: At 3FFB2AC0 len 0002D540 (181 KiB): DRAM
I (261) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM
I (267) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (274) heap_init: At 4008A03C len 00015FC4 (87 KiB): IRAM
I (280) cpu_start: Pro cpu start user code
I (298) spi_flash: detected chip: generic
I (299) spi_flash: flash io: dio
W (299) spi_flash: Detected size(4096k) larger than the size in the binary image header(2048k). Using the size in the binary image header.
I (309) cpu_start: Starting scheduler on PRO CPU.
I (0) cpu_start: Starting scheduler on APP CPU.
E (10320) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (10320) task_wdt:  - loop_task (CPU 0/1)
E (10320) task_wdt: Tasks currently running:
E (10320) task_wdt: CPU 0: IDLE0
E (10320) task_wdt: CPU 1: IDLE1

```

**Comment by david-cermak** [comment]
@will-emmerson 
Ok, thanks for this test, now I think i'm slowly getting what you meant...
Sorry have missed from the beginning that you're actually watching your original `mqtt-publish` (thought when you were talking about WDT default settings, so i was under the impression that only IDLE tasks are being watched).

Is it correct? Are you watching the other tasks as well? In that case, yes, you would have to reset the task WDT at least every `CONFIG_ESP_TASK_WDT_TIMEOUT_S` period. So if the publish function is called with timeout of 10s, the watchdog timeout must be higher.

Now, to the example above: Running the
``` c
    ESP_ERROR_CHECK(esp_task_wdt_add(NULL));
``` 
would register this task to be watched, so the user must reset the WDT (no actual relation to yielding nor task priorities here).  By default (without calling the `esp_task_wdt_add()` function), only IDLE tasks are watched, so no need to reset anything. Just yielding to low-prio task would reset the watchdog indirectly, indicating the tasks with the lowest prio have chance on running.

**Comment by will-emmerson** [comment]
Ok, it's starting to make sense now, sorry for the confusion. Yes I'm watching all tasks with the watchdog because I had a weird problem where one of them was getting 'stuck' and ultimately stopping the whole program. So what you're saying is that if I'm watching my own tasks then I need to be extra careful about what code I'm calling in idf because it might block and it's not going to reset the task watchdog itself?

**Comment by david-cermak** [comment]
> So what you're saying is that if I'm watching my own tasks then I need to be extra careful about what code I'm calling in idf because it might block and it's not going to reset the task watchdog itself?

Exactly! see here https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/wdts.html#task-watchdog-timer
> Each watched task must ‘reset’ the TWDT periodically to indicate that they have been allocated CPU time.

In contrast (with default settings) when watching the idle tasks only, you just have to be careful to yield so lower priorities.

**Comment by will-emmerson** [comment]
Is it worth a note in the docs saying this a bit more obviously? Because with the default task watchdog (5 seconds) and default network timeout (10 seconds) there is going to be problem.

**Comment by david-cermak** [comment]
> Is it worth a note in the docs

I don't think that all the details about network timeouts should be described in the ["task watchdog docs"](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/wdts.html#task-watchdog-timer) (since no task calling a network/blocking API is watched by default).
But agree, that it is not very clear from the MQTT docs, that some APIs might block with certain timeout: 
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/protocols/mqtt.html?highlight=mqtt#_CPPv423esp_mqtt_client_publish24esp_mqtt_client_handle_tPKcPKciii

I suggest adding the notes on blocking APIs and timeouts to the MQTT docs. If you agree with this assessment, please keep this open and will close automatically on adding these docs.
